Terminal.Gui/Views/GraphView.cs

```diff
@@ -0,0 +1,1913 @@
+				driver.AddRune (driver.TopTee);
+
+				// and the label text
+				if (!string.IsNullOrWhiteSpace (label.Text)) {
+
+					// center the label but don't draw it outside bounds of the graph
+					int drawAtX = Math.Max (0, label.ScreenLocation.X - (label.Text.Length / 2));
+					string toRender = label.Text;
+
+					// this is how much space is left
+					int xSpaceAvailable = graph.Bounds.Width - drawAtX;
+
+					// There is no space for the label at all!
+					if (xSpaceAvailable <= 0) {
+						continue;
+					}
+
+					// if we are close to right side of graph, don't overspill
+					if (toRender.Length > xSpaceAvailable) {
+						toRender = toRender.Substring (0, xSpaceAvailable);
+					}
+
+					graph.Move (drawAtX, Math.Min (label.ScreenLocation.Y + 1, bounds.Height - 1));
+					driver.AddStr (toRender);
+				}
+			}
+
+			// if there is a title
+			if (!string.IsNullOrWhiteSpace (Text)) {
+
+				string toRender = Text;
+
+				// if label is too long
+				if (toRender.Length > graph.Bounds.Width) {
+					toRender = toRender.Substring (0, graph.Bounds.Width);
+				}
+
+				graph.Move (graph.Bounds.Width / 2 - (toRender.Length / 2), graph.Bounds.Height - 1);
+				driver.AddStr (toRender);
+			}
+		}
+
+		private IEnumerable<AxisIncrementToRender> GetLabels (GraphView graph, Rect bounds)
+		{
+			// if no labels
+			if (Increment == 0) {
+				yield break;
+			}
+
+			int labels = 0;
+			int y = GetAxisYPosition (graph, bounds);
+
+			for (int i = 0; i < bounds.Width; i++) {
+
+				// what bit of the graph is supposed to go here?
+				var graphSpace = graph.ScreenToGraphSpace (i, y);
+
+				if (Math.Abs (graphSpace.X) % Increment < graph.CellSize.X) {
+					var toRender = new AxisIncrementToRender (Orientation, new Point (i, y), graphSpace);
+
+					if (ShowLabelsEvery != 0) {
+
+						// if this increment also needs a label
+						if (labels++ % ShowLabelsEvery == 0) {
+							toRender.Text = LabelGetter (toRender);
+						};
+					}
+
+					yield return toRender;
+				}
+			}
+		}
+		/// <summary>
+		/// Returns the Y screen position of the origin (typically 0,0) of graph space.
+		/// Return value is bounded by the screen i.e. the axis is always rendered even
+		/// if the origin is offscreen.
+		/// </summary>
+		/// <param name="graph"></param>
+		/// <param name="bounds"></param>
+		private int GetAxisYPosition (GraphView graph, Rect bounds)
+		{
+			// find the origin of the graph in screen space (this allows for 'crosshair' style
+			// graphs where positive and negative numbers visible
+			var origin = graph.GraphSpaceToScreen (new PointD (0, 0));
+
+			// float the X axis so that it accurately represents the origin of the graph
+			// but anchor it to top/bottom if the origin is offscreen
+			return Math.Min (Math.Max (0, origin.Y), bounds.Height - ((int)graph.MarginBottom + 1));
+		}
+	}
+
+	/// <summary>
+	/// The vertical (i.e. Y axis) of a <see cref="GraphView"/>
+	/// </summary>
+	public class VerticalAxis : Axis {
```

Comment:
Another example of a building-block class that should be constrainted to GraphView somehow.